crates/vm/levm/src/transaction.rs

```diff
@@ -68,4 +80,254 @@ pub struct TxEnv {
+            if self.blob_hashes.is_empty() {
+                return Err(InvalidTx::EmptyBlobs);
+            }
+
+            // the maximum number of blobs in a block for now is 6
+            // https://eips.ethereum.org/EIPS/eip-4844#throughput
+            if self.number_of_blobs() > MAX_BLOB_NUMBER_PER_BLOCK as u64 {
+                return Err(InvalidTx::TooManyBlobs {
+                    have: self.number_of_blobs() as usize,
+                    max: MAX_BLOB_NUMBER_PER_BLOCK as usize,
+                });
+            }
+
+            // each blob's first byte must be `VERSIONED_HASH_VERSION_KZG` (0x01)
+            // https://github.com/ethereum/execution-specs/blob/c854868f4abf2ab0c3e8790d4c40607e0d251147/src/ethereum/cancun/fork.py#L408
+            for blob in self.blob_hashes.iter() {
+                if blob[0] != VERSIONED_HASH_VERSION_KZG {
+                    return Err(InvalidTx::BlobVersionNotSupported);
+                }
+            }
+
+            // check that the tx is willing to pay at least the blob gas price
+            // https://github.com/ethereum/execution-specs/blob/c854868f4abf2ab0c3e8790d4c40607e0d251147/src/ethereum/cancun/fork.py#L412
+            let excess_blob_gas = block_env
+                .excess_blob_gas
+                .expect("it's a blob tx, but the block has no blob gas price");
+            // TODO: this should probably return an error, but it would violate
+            // revms api.
+
+            let blob_gasprice = Self::calculate_blob_gas_price(excess_blob_gas);
+
+            if max_fee_per_blob_gas < U256::from(blob_gasprice) {
+                return Err(InvalidTx::BlobGasPriceGreaterThanMax);
+            }
+
+            max_gas_fee += self.calculate_total_blob_gas() * max_fee_per_blob_gas;
+        } else if !self.blob_hashes.is_empty() {
+            // if not a blob tx, but there are blob hashes, it's an error
+            return Err(InvalidTx::BlobVersionedHashesNotSupported);
+        }
+
+        let fee = max_gas_fee
+            .checked_add(self.value)
+            .ok_or(InvalidTx::OverflowPaymentInTransaction)?;
+
+        if fee > account.balance {
+            return Err(InvalidTx::LackOfFundForMaxFee {
+                fee: Box::new(fee),
+                balance: Box::new(account.balance),
+            });
+        }
+
+        // TODO: check if more validations are needed
+        Ok(())
+    }
+
+    pub fn validate_transaction(
+        &mut self,
+        account: &Account,
+        block_env: &BlockEnv,
+    ) -> Result<u64, VMError> {
+        let initial_gas_consumed = self
+            .consume_intrinsic_cost()
+            .map_err(|_| VMError::InvalidTransaction)?;
+        self.validate_tx_env(account, block_env)
+            .map_err(|_| VMError::InvalidTransaction)?;
+
+        Ok(initial_gas_consumed)
+    }
+
+    pub fn number_of_blobs(&self) -> u64 {
+        self.blob_hashes.len() as u64
+    }
```

Comment:
Couldn't we return `usize` and avoid the `as` conversion? If not, leave this as it is, we'll handle the `as` conventions later.

Reply:
Done!